test/functional/test_framework/test_node.py

```
@@ -322,13 +322,18 @@ def wait_for_rpc_connection(self, *, wait_for_import=True):
                 suppressed_errors[f"JSONRPCException {e.error['code']}"] += 1
                 latest_error = repr(e)
             except OSError as e:
+                error_num = e.errno
+                # Workaround issue observed on Windows, Python v3.13.1 where socket timeouts don't have errno set.
+                if error_num is None:
+                    assert isinstance(e, TimeoutError)
+                    error_num = errno.ETIMEDOUT
```

The inclusion of an assert suggests to me that we expect that all OSErrors without errno are TimeoutErrors? Is this actually what we expect?

If not should the if statement be changed to `if error_num is None and isinstance(e, TimeoutError):`

---

Had a feeling `OSError`s must always have that field set (except for the socket `TimeoutError` bug), but the [docs](https://docs.python.org/3/library/exceptions.html#OSError) indicate it's a bit more loose:

> The second form of the constructor sets the corresponding attributes, described below. The attributes default to None if not specified.

Could change the condition as you suggestion, but it would be prudent to assert that `error_num` isn't `None` for other cases:
```python
                if error_num is None and isinstance(e, TimeoutError):
                    error_num = errno.ETIMEDOUT
                else:
                    assert error_num is not None
```
So not sure about that change as it would repeat one of the conditions. Maybe just adding an error message to what we have now would be okay?
```python
                if error_num is None:
                    assert isinstance(e, TimeoutError), "We only currently support " \
                        f"inferring errno from TimeoutErrors, encountered unhandled: {e!r}"
                    error_num = errno.ETIMEDOUT
```

---

Not so concerned about people running in optimised node since 5+ functional tests fail in that way, but changed in latest push to:
https://github.com/bitcoin/bitcoin/blob/cac218dcbdd2b5682a7ac8659aac0306e68a65f5/test/functional/test_framework/test_node.py#L325-L331

---

Thanks for re-reviewing this again!

Yes, in that way we would make `error_num == None` be one of the cases for which we `raise`. It's obviously better to `raise` for something unexpected than to suppress, which should lead to it being tracked down and intentionally categorized for suppression or not.

Updated in latest push.